File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/service/AVTransportController.java
package com.android.cast.dlna.dmr.service;


import android.content.Context;

import com.android.cast.dlna.core.Utils;
import com.android.cast.dlna.dmr.DLNARendererActivity;
import com.android.cast.dlna.dmr.IDLNARenderControl;

import org.fourthline.cling.model.types.ErrorCode;
import org.fourthline.cling.model.types.UnsignedIntegerFourBytes;
import org.fourthline.cling.support.avtransport.AVTransportErrorCode;
import org.fourthline.cling.support.avtransport.AVTransportException;
import org.fourthline.cling.support.model.DeviceCapabilities;
import org.fourthline.cling.support.model.MediaInfo;
import org.fourthline.cling.support.model.PositionInfo;
import org.fourthline.cling.support.model.SeekMode;
import org.fourthline.cling.support.model.StorageMedium;
import org.fourthline.cling.support.model.TransportAction;
import org.fourthline.cling.support.model.TransportInfo;
import org.fourthline.cling.support.model.TransportSettings;

import java.net.URI;

public class AVTransportController implements IRendererInterface.IAVTransportControl {
    private static final TransportAction[] TRANSPORT_ACTION_STOPPED = new TransportAction[]{TransportAction.Play};
    private static final TransportAction[] TRANSPORT_ACTION_PLAYING = new TransportAction[]{TransportAction.Stop, TransportAction.Pause, TransportAction.Seek};
    private static final TransportAction[] TRANSPORT_ACTION_PAUSE_PLAYBACK = new TransportAction[]{TransportAction.Play, TransportAction.Seek, TransportAction.Stop};

    private final UnsignedIntegerFourBytes mInstanceId;
    private final Context mApplicationContext;
    private final TransportInfo mTransportInfo = new TransportInfo();
    private final TransportSettings mTransportSettings = new TransportSettings();
    private PositionInfo mPositionInfo = new PositionInfo();
    private MediaInfo mMediaInfo = new MediaInfo();
    private final IDLNARenderControl mMediaControl;

    public AVTransportController(Context context, IDLNARenderControl control) {
        this(context, new UnsignedIntegerFourBytes(0), control);
    }

    public AVTransportController(Context context, UnsignedIntegerFourBytes instanceId, IDLNARenderControl control) {
        mApplicationContext = context.getApplicationContext();
        mInstanceId = instanceId;
        mMediaControl = control;
    }

    public UnsignedIntegerFourBytes getInstanceId() {
        return mInstanceId;
    }

    public synchronized TransportAction[] getCurrentTransportActions() {
        switch (mTransportInfo.getCurrentTransportState()) {
            case PLAYING:
                return TRANSPORT_ACTION_PLAYING;
            case PAUSED_PLAYBACK:
                return TRANSPORT_ACTION_PAUSE_PLAYBACK;
            default:
                return TRANSPORT_ACTION_STOPPED;
        }
    }

    @Override
    public DeviceCapabilities getDeviceCapabilities() {
        return new DeviceCapabilities(new StorageMedium[]{StorageMedium.NETWORK});
    }

    @Override
    public MediaInfo getMediaInfo() {
        return mMediaInfo;
    }

    @Override
    public PositionInfo getPositionInfo() {
        return mPositionInfo;
    }

    @Override
    public TransportInfo getTransportInfo() {
        return mTransportInfo;
    }

    @Override
    public TransportSettings getTransportSettings() {
        return mTransportSettings;
    }

    @Override
    public void setAVTransportURI(String currentURI, String currentURIMetaData) throws AVTransportException {
        try {
            new URI(currentURI);
        } catch (Exception ex) {
            throw new AVTransportException(ErrorCode.INVALID_ARGS, "CurrentURI can not be null or malformed");
        }
        mMediaInfo = new MediaInfo(currentURI, currentURIMetaData, new UnsignedIntegerFourBytes(1), "", StorageMedium.NETWORK);
        mPositionInfo = new PositionInfo(1, currentURIMetaData, currentURI);
        DLNARendererActivity.startActivity(mApplicationContext, currentURI);
    }

    @Override
    public void setNextAVTransportURI(String nextURI, String nextURIMetaData) {
    }

    @Override
    public void play(String speed) {
        mMediaControl.play();
    }

    public void pause() {
        mMediaControl.pause();
    }

    @Override
    public void seek(String unit, String target) throws AVTransportException {
        SeekMode seekMode = SeekMode.valueOrExceptionOf(unit);
        if (!seekMode.equals(SeekMode.REL_TIME)) {
            throw new AVTransportException(AVTransportErrorCode.SEEKMODE_NOT_SUPPORTED, "Unsupported seek mode: " + unit);
        }
        long position = Utils.getIntTime(target);
        mMediaControl.seek(position);
    }

    synchronized public void stop() {
        mMediaControl.stop();
    }

    @Override
    public void previous() {
    }

    @Override
    public void next() {
    }

    @Override
    public void record() {
    }

    @Override
    public void setPlayMode(String newPlayMode) {
    }

    @Override
    public void setRecordQualityMode(String newRecordQualityMode) {
    }
}

File: dlna-dmr/src/main/java/com/android/cast/dlna/dmr/service/AVTransportController.kt
package com.android.cast.dlna.dmr.service

import android.content.Context
import com.android.cast.dlna.core.Logger
import com.android.cast.dlna.dmr.RenderControl
import org.fourthline.cling.model.ModelUtil
import org.fourthline.cling.support.avtransport.AVTransportException
import org.fourthline.cling.support.model.*
import org.fourthline.cling.support.model.TransportAction.*

class AVTransportController(override val applicationContext: Context) : AvTransportControl {
    companion object {
        private val TRANSPORT_ACTION_STOPPED = arrayOf(Play)
        private val TRANSPORT_ACTION_PLAYING = arrayOf(Stop, Pause, Seek)
        private val TRANSPORT_ACTION_PAUSE_PLAYBACK = arrayOf(Play, Seek, Stop)
    }

    var mediaControl: RenderControl? = null
        set(value) {
            if (value != null) {
                _mediaInfo = MediaInfo(currentURI, currentURIMetaData)
                _positionInfo = PositionInfo(0, currentURIMetaData, currentURI)
            } else {
                mediaControl?.stop()
                _mediaInfo = MediaInfo()
                _positionInfo = PositionInfo()
            }
            field = value
        }
    private var _positionInfo = PositionInfo()
    private var _mediaInfo = MediaInfo()

    override val logger = Logger.create("AVTransportController")
    override val transportSettings = TransportSettings()
    override val deviceCapabilities: DeviceCapabilities = DeviceCapabilities(arrayOf(StorageMedium.UNKNOWN))
    override val transportInfo
        get() = mediaControl?.let { ctrl ->
            TransportInfo(ctrl.getState().toTransportState(), TransportStatus.OK, "1")
        } ?: TransportInfo()
    override val mediaInfo
        get() = _mediaInfo
    override val positionInfo: PositionInfo
        get() = mediaControl?.let { ctrl ->
            val duration = ModelUtil.toTimeString(ctrl.duration / 1000)
            val realTime = ModelUtil.toTimeString(ctrl.currentPosition / 1000)
            PositionInfo(0, duration, currentURI, realTime, realTime)
        } ?: PositionInfo()
    override val currentTransportActions: Array<TransportAction>
        get() = when (transportInfo.currentTransportState) {
            TransportState.PLAYING -> TRANSPORT_ACTION_PLAYING
            TransportState.PAUSED_PLAYBACK -> TRANSPORT_ACTION_PAUSE_PLAYBACK
            else -> TRANSPORT_ACTION_STOPPED
        }

    private var currentURI: String? = null
    private var currentURIMetaData: String? = null

    @Throws(AVTransportException::class)
    override fun setAVTransportURI(currentURI: String, currentURIMetaData: String?) {
        super.setAVTransportURI(currentURI, currentURIMetaData)
        this.currentURI = currentURI
        this.currentURIMetaData = currentURIMetaData
    }

    private var nextURI: String? = null
    private var nextURIMetaData: String? = null

    override fun setNextAVTransportURI(nextURI: String, nextURIMetaData: String?) {
        super.setNextAVTransportURI(nextURI, nextURIMetaData)
        this.nextURI = nextURI
        this.nextURIMetaData = nextURIMetaData
    }

    override fun play(speed: String?) {
        super.play(speed)
        mediaControl?.play()
    }

    override fun pause() {
        super.pause()
        mediaControl?.pause()
    }

    override fun seek(unit: String?, target: String?) {
        super.seek(unit, target)
        try {
            mediaControl?.seek(ModelUtil.fromTimeString(target) * 1000)
        } catch (e: Exception) {
            logger.w("seek failed: $e")
        }
    }

    override fun next() {
        super.next()
        if (nextURI != null && nextURIMetaData != null) {
            previousURI = currentURI
            previousURIMetaData = currentURIMetaData
            setAVTransportURI(nextURI!!, nextURIMetaData)
        }
        nextURI = null
        nextURIMetaData = null
    }

    private var previousURI: String? = null
    private var previousURIMetaData: String? = null
    override fun previous() {
        super.previous()
        if (previousURI != null && previousURIMetaData != null) {
            nextURI = currentURI
            nextURIMetaData = currentURIMetaData
            setAVTransportURI(previousURI!!, previousURIMetaData)
        }
        previousURI = null
        previousURIMetaData = null
    }

    override fun stop() {
        super.stop()
        mediaControl?.stop()
        _mediaInfo = MediaInfo()
        _positionInfo = PositionInfo()
    }
}
